Add new model features to capture parcel shape

dfsnow opened the pull request:
This PR updates the parcel file ETL to add 6 new features for modeling. These features attempt to capture the complexity of the parcel shape, based on the idea that more "regular" shaped parcels are more desirable and are thus worth more, they were mostly taken from [this paper](https://eres.architexturez.net/system/files/P_20230206101430_7873.pdf). The 6 features are:

1. The standard deviation of the interior angles of the parcel polygon
2. The standard deviation of the parcel polygon edge length
3. The standard deviation of the parcel vertices' distance to the parcel centroid
4. The total number of vertices in the parcel polygon
5. The ratio of the length of the sides of the minimum spanning rotated rectangle
6. The ratio of the areas of the minimum spanning rotated rectangle and the parcel polygon itself

I haven't actually run this script yet to replace any cleaned up prod parcel files, but I'll do so once this is merged.

I'll make a follow-up PR to add the output features to dbt and all the necessary views/docs.

dfsnow commented on etl/etl.Rproj:
```diff
@@ -12,6 +12,6 @@ Encoding: UTF-8
 
-StripTrailingWhitespace: Yes
 AutoAppendNewline: Yes
+StripTrailingWhitespace: Yes
```
This was auto-formatted by RStudio, not an actual change.

jeancochrane commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -174,29 +194,137 @@ process_parcel_file <- function(s3_bucket_uri,
+      by = "L3"
+    ]
+
+    # Collapse the edge-level data back to the parcel level, this time keeping
+    # the 180 degree vertices, since excluding them would drop some edges with
+    # non-zero length
```
**[Question, non-blocking]** Would it be slightly more accurate to sum the length of any edges that are joined with angle ~180° when executing operations based on edge length, since the additional vertex is most likely a data error cutting a single edge in two? I don't feel super strongly about this (especially if the universe of 180° vertices is small) but I figure it's worth checking!

jeancochrane reviewed (approved):
Very nice work! I don't feel super qualified to evaluate the math or `data.tables` code in this PR, so let's get Billy's thoughts on this too, but overall it makes sense to me. There are a few linter warnings that you can see if you navigate to the PR file browser, might be nice to clean those up before merging but not required.

wrridgeway commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -173,28 +210,159 @@ process_parcel_file <- function(s3_bucket_uri,
+      # Calculate the interior angle between each pair of points
+      angle := calculate_angles(as.matrix(.SD)),
+      .SDcols = c("X", "Y"),
+      by = c("L1", "L2", "L3")
```
I'm having a hard time parsing what L1, L2, and L3 are?

dfsnow replied:
They are one of [the outputs](https://r-spatial.github.io/sf/reference/st_coordinates.html#value) from `st_coordinates()`. Basically, L3 corresponds to the PIN10 group while L1 and L2 can represent different sub-polygons of a multi-polygon.

dfsnow commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -122,40 +148,67 @@ process_parcel_file <- function(s3_bucket_uri,
       group_by(pin10) %>%
       arrange(desc(area)) %>%
-      summarize(across(c(lon, lat, x_3435, y_3435), first)) %>%
+      summarize(across(c(lon, lat, x_3435, y_3435), dplyr::first)) %>%
```
Namespacing here just to prevent conflicts with `data.table`, which also has a `first()` function.

dfsnow commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -122,40 +148,67 @@ process_parcel_file <- function(s3_bucket_uri,
-        geometry_3435 = st_transform(geometry, 3435)
+        area = units::drop_units(st_area(geometry))
       ) %>%
+      # Drop anything with an area less than 1 sq meter, as long as it has other
+      # polygons for the same PIN10 that are larger
+      filter(!(area <= 1 & n() >= 2 & !all(area <= 1)), .by = "pin10") %>%
```
There were ~200 parcels that had "sliver" polygons that were messing up the feature results. Here we drop these slivers i.f.f. they belong to the same PIN10 as a larger polygon. This cleans up the feature results significantly.

dfsnow commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -122,40 +148,67 @@ process_parcel_file <- function(s3_bucket_uri,
-        lon, lat, x_3435, y_3435, geometry, geometry_3435,
+        lon, lat, x_3435, y_3435, geometry,
         town_code, year
-      ) %>%
-      distinct(pin10, .keep_all = TRUE)
```
This was actually a mistake/bug, so I'm glad I reviewed the rest of the parcel cleaning code. This `distinct()` drops an arbitrary polygon in cases where there are multiple polygons per PIN10, definitely not what we want. It wouldn't have actually impacted any model results or other processes, since the centroids (used for modeling) are (correctly) calculated separate from this block of code.

dfsnow commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -122,40 +148,67 @@ process_parcel_file <- function(s3_bucket_uri,
+      )
+    tictoc::toc()
+
+    # Aggregate PIN10s with more than one row using st_union. Split them out
+    # from the rest of the dataset because st_union is an incredibly expensive
+    # operation, so we only want to perform it on the PINs we need to
+    tictoc::tic(paste("Collapsed parcels to the PIN10 level:", file_year))
+    spatial_df_merged <- bind_rows(
+      spatial_df_merged %>%
+        filter(n() >= 2, .by = pin10) %>%
+        summarize(
+          across(tax_code:year, dplyr::first),
+          across(geometry, st_union),
+          .by = "pin10"
+        ),
+      spatial_df_merged %>%
+        filter(!(n() >= 2), .by = pin10)
+    ) %>%
+      st_cast("MULTIPOLYGON") %>%
+      mutate(geometry_3435 = st_transform(geometry, 3435)) %>%
+      relocate(geometry_3435, .after = geometry)
+    tictoc::toc()
```
Here we aggregate to the PIN10 level using a geometry union instead of `distinct()`. This also improves the feature results quite a bit, as they now include all polygons.

dfsnow commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -173,28 +226,159 @@ process_parcel_file <- function(s3_bucket_uri,
+    # Using data.table here because it's much faster than dplyr
+    spatial_mat_coords <- spatial_df_merged %>%
+      st_set_geometry("geometry_3435") %>%
+      st_simplify(dTolerance = 2, preserveTopology = TRUE) %>%
+      st_cast("MULTIPOLYGON") %>%
```
Added a simplification step with a low, low tolerance (2 feet) to remove points that are basically overlapping.

dfsnow commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -173,28 +226,162 @@ process_parcel_file <- function(s3_bucket_uri,
+      by = c("L1", "L2", "L3")
+    ][
+      ,
+      # Some vertex angles are close to 180 or 360 degrees, which are likely to
+      # be artifacts of the data and not true vertices. We don't want to count
+      # these straight angles in our features, BUT we also can't drop all of
+      # them since some polygons are basically just lines. So, we drop any
+      # straight angle vertices from any polygon with >= 3 non-straight angles
+      # (which is the minimum number of vertices needed to make a polygon)
+      straight_angle := (
+        (
+          data.table::between(angle, 0, 1) |
+            data.table::between(angle, 359, 360)
+        ) | data.table::between(angle, 179, 181)
+      )
+    ][
+      ,
+      num_non_straight_angle := sum(!straight_angle, na.rm = TRUE),
+      by = c("L1", "L2", "L3")
+    ][
+      !straight_angle | num_non_straight_angle < 3 | is.na(angle),
```
Per @jeancochrane's prior comment, this calculates the interior vertex angle and removes straight angles _before_ calculating the rest of the features (edge length, distance to centroid, etc.). This yields way more accurate feature results, nearly all the PINs with a high angle SD are wacky things like this:

![f71be41a-18b1-4129-b980-a84ec9997a92](https://github.com/ccao-data/data-architecture/assets/31494343/4e39dedb-a674-4a6c-9b4e-cc682dc8a14a)

dfsnow commented:
@jeancochrane @wrridgeway I'm requesting re-review on this as I just made some pretty significant changes. Mainly:

- Fixed an error that was dropping arbitrary parts of multi-polygon parcels
- Added a simplification step to remove overlapping points
- Added a filtering step to remove polygon slivers and lines
- Improved the feature calculation by first removing vertices with straight angles

All these changes should not only give us more accurate shape features, they should also give us a more accurate/cleaner parcel shapefile.

jeancochrane commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -173,28 +226,162 @@ process_parcel_file <- function(s3_bucket_uri,
+        shp_parcel_mrr_area_ratio = units::drop_units(
+          st_area(geometry_3435) / parcel_area
+        ),
+        # Sometimes the ratio here is very slightly less than 1, even though
+        # that should be impossible. Seems to an artifact of some malformed
```
**[Question, non-blocking]** Not a huge deal, but the last sentence is a fragment and I'm wondering what you think might be malformed here?

dfsnow replied:
It seems to only be the result when calculating the ratio for these super long, skinny triangles, though I'm really not sure why. The minimum rotated rectangle is in red below. It definitely _should_ have an area larger than the triangle, but it is in fact smaller by 0.00657 ft^2. I don't think it's worth debugging further, given the magnitude of the error. Will update my comment to this effect.

![2754c101-1d09-48ad-9c59-4bd5f8bfe59c](https://github.com/ccao-data/data-architecture/assets/31494343/76f6ba3a-dff9-4173-8392-9a9c7a884b93)

jeancochrane reviewed (approved):
Nice work!

dfsnow commented:
@jeancochrane @wrridgeway Okay, last time requesting review on this. I had to make a bunch of edits to get the new features to even run (mostly in 16fcb07). The major changes are:

- Run an `st_make_valid()` call _before_ filtering out invalid parcels. This keeps some parcels that were getting dropped before, resulting in a larger parcel count per year than previously (see below).
- Simplify the geometry processing/transformations so that we're not transforming back and forth between 3435 and 4326. This is not only faster, it also preserves some parcels which are made invalid by the transformation.
- Add a check at the end of the script to ensure that the final count of unique PIN10s is very close to the initial count from the raw parcel file. They don't match exact because some parcels have empty geometries and are dropped.

There's a temporary `spatial.parcel_new` table in Athena if you all want to poke around it. Here's the yearly row counts between the existing table and the new one.

```sql
SELECT *
FROM (
    SELECT year, COUNT(*) AS cnt_old
    FROM spatial.parcel
    GROUP BY year
)
LEFT JOIN (
    SELECT year, COUNT(*) AS cnt_new
    FROM spatial.parcel_new
    GROUP BY year
) USING (year)
ORDER BY year DESC
```

| year | cnt_old | cnt_new | diff |
|------|---------|---------|------|
| 2023 | 1417110 | 1417110 | 0    |
| 2022 | 1416347 | 1416420 | 73   |
| 2021 | 1415410 | 1416038 | 628  |
| 2020 | 1415664 | 1415729 | 65   |
| 2019 | 1415193 | 1415409 | 216  |
| 2018 | 1414731 | 1414794 | 63   |
| 2017 | 1414896 | 1414954 | 58   |
| 2016 | 1414449 | 1414506 | 57   |
| 2015 | 1414155 | 1414227 | 72   |
| 2014 | 1413788 | 1413862 | 74   |
| 2013 | 1413759 | 1413832 | 73   |
| 2012 | 1413732 | 1413809 | 77   |
| 2011 | 1413698 | 1413775 | 77   |
| 2010 | 1413699 | 1413777 | 78   |
| 2009 | 1413740 | 1413816 | 76   |
| 2008 | 1413206 | 1413280 | 74   |
| 2007 | 1410605 | 1410677 | 72   |
| 2006 | 1408738 | 1408808 | 70   |
| 2005 | 1405047 | 1405112 | 65   |
| 2004 | 1401506 | 1401570 | 64   |
| 2003 | 1398092 | 1398094 | 2    |
| 2002 | 1395387 | 1395386 | -1   |
| 2001 | 1393174 | 1393174 | 0    |
| 2000 | 1391620 | 1391620 | 0    |

And here's an example of the type of invalid parcel we were previously dropping:

<img width="652" alt="Screenshot 2024-07-08 at 12 04 28 PM" src="https://github.com/ccao-data/data-architecture/assets/31494343/0258a29e-52f9-4f40-9aee-762a62448a53">

After this is merged we should re-run all the dbt models downstream of `spatial.parcel`, just to be safe.

jeancochrane commented on etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R:
```diff
@@ -173,28 +222,181 @@ process_parcel_file <- function(s3_bucket_uri,
+        .after = "shp_parcel_edge_len_ft_sd"
+      )
+
+    # Check that the final number of distinct, well-formed parcels is close to
+    # the same as the number in the raw parcel file
+    num_parcel_raw <- spatial_df_raw %>%
+      rename_with(tolower) %>%
+      filter(!is.na(pin10), nchar(pin10) == "10", !st_is_empty(geometry)) %>%
+      st_drop_geometry() %>%
+      summarize(n = n_distinct(pin10)) %>%
+      pull(n)
+    num_parcel_final <- spatial_df_final %>%
+      st_drop_geometry() %>%
+      summarize(n = n_distinct(pin10)) %>%
+      pull(n)
+    if (abs(num_parcel_raw - num_parcel_final) > 10) {
+      stop("More than 10 PINs dropped during geometry processing!")
+    }
```
**[Thought, non-blocking]** Good thinking adding this test! It makes me wonder more broadly about how we should define QC tests on non-dbt transformation scripts like this one. Maybe running tests like this at the end of the script is the best solution? Or maybe we split tests out into a function defined in a different module, and then import and call that function at the end of this script for better separation of concerns? Either way, I support the current solution in the immediate term; this is just a thought to keep in mind if we find ourselves adding more QC tests like this one to non-dbt transformations.

dfsnow replied:
Let's leave it for now, since this script run rarely enough that I don't think investing lots of time into them is a good idea. However, this is definitely something to think about when we look at #518.

jeancochrane reviewed (approved):
Really impressive and subtle work in this PR. Thanks for the attention to detail!

wrridgeway commented:
Rad work dude. I didn't get around to checking the math, but everything else looked just great.
